### DISCUSSION TITLE
(無主旨)我是一位前端开发。最近想尝试了解如何制作一个输入法

### DISCUSSION THREAD

**Discussion by lotem** [rime/home#1255] [General]
我是一位前端开发。

最近想尝试了解如何制作一个输入法，然后我查了很多地方，发现讲解如何开发一个输入法的文章或者帖子几乎没有，现在让我很没头绪。在搜索工程中，发现了你们的输入法。

我的一个小小的计划是通过typescript写一个小小的demo，目的是针对自己有想要做一个不是很常见的一些符号的输入法。所以我想找你们了解一些，做一个输入法需要具备的知识和开发的工作流。

**Comment by lotem** [comment]
（現在有了）

通過 web API 不能直接做出一個系統級的輸入法，不過做個 demo 完全可行。

只需在網頁上設置一個 `<input type="text">` 作爲輸入框，偵聽其 [`keydown`](https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event), `keyup` 事件，進而實現按鍵處理邏輯、輸入序列到目標文字的轉換，最終將完成轉換的目標文字添加到文本框的 [`value`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/text#value) 字符串中，即演示了一次輸入法的轉換過程。

Rime 的工作流程大致可以總結爲，用以下 `processKeyEvent` 按鍵處理函數處理接收到的鍵盤事件，將需要輸入法處理的按鍵序列轉換成文字後，通過回調 `commitText` 上屏。

```typescript
enum ProcessResult {
  Rejected,  // 輸入法不處理該事件，並繼續傳送給 input 控件做默認處理
  Accepted,  // 輸入法已處理該事件，不再繼續傳遞
}
// 處理 keydown, keyup 事件
function processKeyEvent(event: Event, commitText: (text: string) => void): ProcessResult {
  if (event.type == 'keydown') {
    if (event.keyCode == 13) {
      commitText("Nuclear launch detected. ");
      return Accepted;
    }
  }
  return Rejected;
}
```

在這個函數裏，可以根據輸入法的功能需求編寫邏輯代碼，通常要處理充當輸入編碼、用於選字的按鍵以及用於干預輸入法狀態的編輯功能鍵。

**Reply by eagleoflqj** [reply]
哈哈哈哈这我可太熟悉了（本回复内容使用My RIME打出
https://my-rime.vercel.app/
